### PR TITLE
v0.0.1

### PR DESCRIPTION
**Pull request by joshuafleck**
For this story: https://trello.com/c/n5dAZN3l/141-3-ab-testing-snowplow-cookie-id-spike

This will allow us to send events from the server side populated with a Snowplow domain_userid, which is a prerequisite for sending AB test start events.

@sparrovv @pvdb @Kimblebrook @martinbtt @danisola  Could you please sign off?

### CODE REVIEW

**Comment by Kimblebrook** [comment]
This all sounds shamazing. For my clarity: will a user still have the same cookie across both the server and client? In other words, if it get sets server side, that exact same cookie will continue to be used by the javascript tracker on the client side?

**Comment by joshuafleck** [comment]
@Kimblebrook 

>if it get sets server side, that exact same cookie will continue to be used by the javascript tracker on the client side?

Yes, that is correct!

**Comment by martinbtt** [comment]
Nice! Looks pretty comprehensive, I've not checked thoroughly against the JS tracker spec so assuming it all hangs together. Two questions:

**1)  ID collision**
It looks like there is potential room for ID collisions? (as in we don't use browser window dimensions etc). The only place I guess this could be an issue is when we have lots of people all coming from the same IP with same HTTP accept/lang features. Which might be more likely for when all of us are testing against staging / integration? 

**2) Release testing**
Whats the plan to release / test? It's a pretty big change and if something is slightly wrong then we could have customers with a long running cookie with incorrect info I guess (worse case). Do we have the end-to-end tests to check that all this goes into Aerie/redshift ok prior to a production release?

**Comment by danisola** [comment]
@martinbtt points are very important. In AERIE we have seen many users with the same fingerprint, mainly users of popular devices (iPhones, iPads, etc). If we can't properly identify a user server-side, maybe we could:

1) Use [SnowplowRubyDuid features](https://github.com/simplybusiness/snowplow_ruby_duid/blob/v0_0_1/lib/snowplow_ruby_duid/domain_userid.rb#L5) as a "temporary identifier" (valid for 5-10min?) and then update our ID's and send events when we get the "real" Snowplow ID in the 2nd request. This solution is quite complex and we should think about what to do in case there is no 2nd request.

2) Or we could just send a totally random ID. This would fail to identify the same user if he/she wipes the cookies, but at least we would avoid collisions.

Also, I'm very keen on having this thoroughly tested. JS tracker bugs are the worst ones we can have!

File: spec/lib/snowplow_ruby_duid/domain_userid.feature [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+  Background:
+    Given a request environment
+      And the time is '2015-01-22 15:26:31 +0000'
+
+  Scenario: The domain userid is 16 characters long
+    When I create a domain userid
+    Then the domain userid has a length of '16'
+
+  Scenario Outline: The time is a component of the domain userid fingerprint
```
Comment (sparrovv) [inline spec/lib/snowplow_ruby_duid/domain_userid.feature]: Is that in JS tracker spec as well? Or we just want to reduce probability of conflicts?

**Comment by sparrovv** [comment]
@danisola I just read the `domain_userid.feature` and looks like `time` is part of the domain userid fingerprint. If that's how it should work, then I'm not really sure how we could identify visitor after they remove the `snowplow` cookie.

**Comment by danisola** [comment]
True! I missed the [request_created_at](https://github.com/simplybusiness/snowplow_ruby_duid/blob/v0_0_1/lib/snowplow_ruby_duid/domain_userid.rb#L23) part. If we're using time maybe we could just use a random value?

We'll have to agree that not identifying users in case they remove cookies is alright though. What are your thoughts on that @Kimblebrook ?

**Comment by Kimblebrook** [comment]
@danisola I am getting a little confused between the `user_fingerprint` and the `domain_userid`. When you say fingerprint, are talking about the cookie or the fingerprinting algorithm that Snowplow uses based on browser settings? My understanding is that these two things are completely separate. 

_> We'll have to agree that not identifying users in case they remove cookies is alright though_

At the moment, when a user deletes cookies, we lose the ability to track them across sessions. I assume this will still be the case. I guess my question then is: given the proposal above, will we potentially be worse off (in terms of losing the ability to link customers) than the status quo?

**Comment by danisola** [comment]
It is quite confusing indeed. The snowplow guys use a set of [browser features and a timestamp](https://github.com/snowplow/snowplow-javascript-tracker/blob/d3d10067127eb5c95d0054c8ae60f3bdccba619d/src/js/tracker.js#L468-L472) to create a random `domain_userid`. Why, I don't know. As far as I understand, any random value would work without need to use browser features.

So my take is that we can generate a random value that has the same format at Snowplow's without having to emulate the way they do it. I think that would simplify the code quite a bit.

> I guess my question then is: given the proposal above, will we potentially be worse off (in terms of losing the ability to link customers) than the status quo?

No, we would be in the same position. I was mistaken by the use of browser features.

File: spec/lib/snowplow_ruby_duid/domain_userid.feature (hunk 1 above)
Comment (joshuafleck) [inline spec/lib/snowplow_ruby_duid/domain_userid.feature]: Yes, it is. See: https://github.com/snowplow/snowplow-javascript-tracker/blob/d3d10067127eb5c95d0054c8ae60f3bdccba619d/src/js/tracker.js#L468-L472

Comment (sparrovv) [inline spec/lib/snowplow_ruby_duid/domain_userid.feature]: Gotcha, so why this algorithm is better than UUID https://github.com/assaf/uuid ?

**Comment by joshuafleck** [comment]
@danisola @sparrovv - I've modified this to use [SecureRandom](http://ruby-doc.org/stdlib-2.1.2/libdoc/securerandom/rdoc/SecureRandom.html#method-c-hex). Is that better?

**Comment by joshuafleck** [comment]
@martinbtt 

> Do we have the end-to-end tests to check that all this goes into Aerie/redshift ok prior to a production release?

These checks will need to be done manually once this library is utilised in Chopin & Seedy

**Comment by sparrovv** [comment]
> @danisola @sparrovv - I've modified this to use SecureRandom. Is that better?

The implementation is much simpler :+1:

**Comment by danisola** [comment]
@joshuafleck I think it is more understandable now. Like it!

**Pull request merged**